File: source/go/internal/core/panchakarahita.go
package core

import (
	"github.com/ishankgupta95/panchang/source/go/v5/internal/types"
	"github.com/ishankgupta95/panchang/source/go/v5/internal/utils"
)

func ComputePanchakaRahita(
	sunriseUtcMs, nextSunriseUtcMs int64,
	getMoon LongitudeAt,
) []types.UtcWindow {
	inPanchakaAt := func(ms int64) bool { return getMoon(ms) >= 300 }

	startInP := inPanchakaAt(sunriseUtcMs)
	endInP := inPanchakaAt(nextSunriseUtcMs)

	if startInP == endInP {
		if startInP {
			return []types.UtcWindow{}
		}
		return []types.UtcWindow{{StartMs: sunriseUtcMs, EndMs: nextSunriseUtcMs}}
	}

	targetDeg := 300.0
	if startInP {
		targetDeg = 0
	}
	crossing := bisectPanchakaBoundary(sunriseUtcMs, nextSunriseUtcMs, inPanchakaAt, targetDeg, getMoon)
	if startInP {
		return []types.UtcWindow{{StartMs: crossing, EndMs: nextSunriseUtcMs}}
	}
	return []types.UtcWindow{{StartMs: sunriseUtcMs, EndMs: crossing}}
}

func bisectPanchakaBoundary(
	loUtcMs, hiUtcMs int64,
	predicate func(ms int64) bool,
	targetDeg float64,
	getMoon LongitudeAt,
) int64 {
	const bracketMs = 120_000
	const maxIters = 30
	startState := predicate(loUtcMs)
	lo, hi := loUtcMs, hiUtcMs

	for i := 0; i < maxIters && hi-lo > bracketMs; i++ {
		mid := floorDiv2(lo + hi)
		if predicate(mid) == startState {
			lo = mid
		} else {
			hi = mid
		}
	}
	if solved, ok := utils.SolveAngleCrossing(float64(lo), float64(hi), targetDeg, getMoon,
		func(ms int64) bool { return predicate(ms) == startState }); ok {
		return solved
	}
	return hi
}

func floorDiv2(n int64) int64 {
	q := n / 2
	if n%2 != 0 && n < 0 {
		q--
	}
	return q
}

File: source/ts/src/core/panchakaRahita.ts
import { solveAngleCrossing } from '../utils/search';
import type { UtcWindow } from '../types/elements';
import { validateDate } from '../utils/validation';

/** Half-open `[start, end)` windows of the Hindu day with the Moon OUTSIDE Panchaka, i.e. sidereal longitude in [0°, 300°). */
export function computePanchakaRahita(
  sunriseUtc: Date,
  nextSunriseUtc: Date,
  getMoon: (d: Date) => number,
): UtcWindow[] {
  validateDate(sunriseUtc, 'any');
  validateDate(nextSunriseUtc, 'any');
  const inPanchakaAt = (d: Date) => getMoon(d) >= 300;

  const startInP = inPanchakaAt(sunriseUtc);
  const endInP = inPanchakaAt(nextSunriseUtc);

  if (startInP === endInP) {
    return startInP ? [] : [{ start: sunriseUtc, end: nextSunriseUtc }];
  }

  const targetDeg = startInP ? 0 : 300;
  const crossing = bisectBoundary(sunriseUtc, nextSunriseUtc, inPanchakaAt, targetDeg, getMoon);
  return startInP
    ? [{ start: crossing, end: nextSunriseUtc }]
    : [{ start: sunriseUtc, end: crossing }];
}

/** Bisection only narrows the bracket; {@link solveAngleCrossing} finishes, so the
 *  answer is not quantized to `BRACKET_MS`. */
function bisectBoundary(
  loUtc: Date,
  hiUtc: Date,
  predicate: (d: Date) => boolean,
  targetDeg: number,
  getMoon: (d: Date) => number,
): Date {
  const BRACKET_MS = 120_000;
  const MAX_ITERS = 30;
  const startState = predicate(loUtc);
  let lo = loUtc.getTime();
  let hi = hiUtc.getTime();

  for (let i = 0; i < MAX_ITERS && hi - lo > BRACKET_MS; i++) {
    const mid = Math.floor((lo + hi) / 2);
    if (predicate(new Date(mid)) === startState) lo = mid;
    else hi = mid;
  }
  const solved = solveAngleCrossing(
    lo, hi, targetDeg, getMoon, (ms) => predicate(new Date(ms)) === startState,
  );
  return new Date(solved ?? hi);
}
